ziprofs.py

```diff
@@ -161,24 +131,39 @@ def getattr(self, path, fh=None):
         return result
 
     def open(self, path, flags):
-        zip_path = self.get_zip_path(path)
-        if not zip_path:
-            return os.open(path, flags)
-        return 0
+        if zip_path := self.get_zip_path(path):
+            fh = self._get_free_zip_fh()
+            zf = self.zip_factory.get(zip_path)
+            self._zip_file_fh[fh] = zf.open(path[len(zip_path) + 1:])
+            self._zip_lock_fh[fh] = Lock()
+            return fh
+        else:
+            return os.open(path, flags) << 1
 
     def read(self, path, size, offset, fh):
-        with self.rwlock:
-            zip_path = self.get_zip_path(path)
-            if not zip_path:
-                os.lseek(fh, offset, 0)
-                return os.read(fh, size)
-            zf = self.zip_factory.get(zip_path)
-            subpath = path[len(zip_path)+1:]
-            return zf.read(subpath, size, offset)
+        if self.get_zip_path(path):
+            f = self._zip_file_fh[fh]  # should be here (file is first opened, then read)
+            with self._zip_lock_fh[fh]:
+                if f.seekable():
+                    foffset = f.tell()
+                    self.log.debug(f" file offset: {foffset}, read offset: {offset}, diff: {offset-foffset}")
+                    f.seek(offset)
+                    return f.read(size)
+                else:
+                    # emulate seek by reading and discarding chunks
+                    filepos = f._orig_file_size - f._left - len(f._readbuffer) + f._offset      # zipfile.py#1110 - tell()
+                    offset -= filepos
+                    if offset < 0:
+                        raise FuseOSError(errno.ENOTSUP)
```

Comment:
I don't think you can return ENOTSUP on read calls. When you are asked to backseek on a non-seekable file reopen it and seek from start.

Reply:
If file is non-seekable every next try to open will fail, only non-seekable example of ZipExtFile is when underlying ZipFile is unseekable.
I could replace this with invalid argument error. What do you think?

Reply:
https://manpage.me/index.cgi?apropos=0&q=read&sektion=2&manpath=Debian+8.1.0&arch=default&format=html#end
You are right, I can't raise ENOSUP.